packages/eslint-plugin/src/rules/no-import-type-side-effects.ts

```diff
@@ -29,6 +29,10 @@ export default util.createRule<Options, MessageIds>({
       'ImportDeclaration[importKind!="type"]'(
         node: TSESTree.ImportDeclaration,
       ): void {
+        if (node.specifiers.length === 0) {
+          return;
+        }
```

Comment:
Per our recent conversation somewhere, if you wanted to, ESQuery can reduce this code count a bit 😄 

```suggestion
      'ImportDeclaration[importKind!="type"][specifiers.0]'(
        node: TSESTree.ImportDeclaration,
      ): void {
```

Reply:
back in the day I was all in on esquery selectors.
now a days though I tend to find that leaning too heavily into the custom DSL makes code less accessible and harder to understand / debug.

`[specifiers.0]` is one of those parts of the DSL that's super unclear as to what it's doing, IMO.

---

also as an interesting aside just because...
I figured out that selectors actually hide performance cost of rules! ESLint only tracks (1) time spent in `create` and (2) time spent in selector functions for a rule.

so time spent compiling and evaluation selector strings is perf cost in ESLint land - not for the rule.

Which is funny for rules like those in [eslint-plugin-unicorn](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/2a9c840484ab456ef131cb7ffc61cd46bef443a7/rules/no-useless-length-check.js#L11-L33) which heavily (ab)use selectors over logic, because they'll look like they're faster than other rules!

Reply:
Ooh I wonder if we should have a lint rule for selectors. And maybe put it in `eslint-plugin-eslint-plugin`...

Reply:
Separately I've also envisioned a lint rule that enforces the type annotation on the function arg matches the selector to improve safety.

EG
```
// bad
'ImportDeclaration[importKind != "type"]'(node: TSESTree.Literal)
'BlockStatement > *'(node: TSESTree.Literal)

// good
'ImportDeclaration[importKind != "type"]'(node: TSESTree.ImportDeclaration)
'ImportDeclaration[importKind != "type"]'(node: TSESTree.ImportDeclaration & {importKind: 'type'})
'BlockStatement > *'(node: TSESTree.Node)
```

Reply:
* https://github.com/eslint-community/eslint-plugin-eslint-plugin/issues/339
* https://github.com/eslint-community/eslint-plugin-eslint-plugin/issues/340